### PR TITLE
[Docs] Added documentation on pytest target parametrization.

### PR DESCRIPTION
**Pull request by Lunderberg**
Follow-up from #8542, to document existing features.

### CODE REVIEW

**Comment by Lunderberg** [comment]
@mbrookhart The documentation that we had discussed, describing how tests can be run across multiple targets.

@hogepodge I have this in the `dev` section of the documentation, trying to follow your proposed re-organization of the documentation.

File: docs/dev/pytest_target_parametrization.rst [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+.. _pytest-marks: https://docs.pytest.org/en/6.2.x/mark.html
+
+Each test file has one or more test functions.  The test functions can
+be decorated with `pytest marks <pytest-marks>`_ to include metadata.
+The most frequently applied marks are as follows.
+
+- ``@pytest.mark.gpu`` - Tags a function as using GPU
+  capabilities. This has no effect on its own, but can be paired with
+  command-line arguments ``-m gpu`` or ``-m 'not gpu'`` to restrict
+  which tests pytest will executed. Typically not called on its own.
```
Comment (tkonolige) [inline docs/dev/pytest_target_parametrization.rst:54]: I'd specify that developers should never use this mark explicitly.

File: docs/dev/pytest_target_parametrization.rst [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+- ``@tvm.testing.uses_gpu`` - Applies ``@pytest.mark.gpu``.  Needed
+  only for tests that explicitly loop over
```
Comment (tkonolige) [inline docs/dev/pytest_target_parametrization.rst:57]: Suggest that this is what should be used to mark tests requiring GPUs, but prefer parameterize_targets.

File: docs/dev/pytest_target_parametrization.rst [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+hardware present.  Most current tests use this style, but it is not
+recommended for new tests.  The pytest output for this style silently
+skips runtimes that are disabled in ``config.cmake``, or do not have a
+device on which they can run.  In addition, the test halts on the
+first target to fail, which is ambiguous as to whether the error
+occurs on a particular target, or on every target.
+
+.. code-block:: python
+
+    # Old style, not recommend anymore
```
Comment (tkonolige) [inline docs/dev/pytest_target_parametrization.rst:83]: I'd prefer saying "do not use" vs "not recommended".

File: docs/dev/pytest_target_parametrization.rst [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+For tests that should run correctly on all targets, the decorator can
+be omitted.  Any test that accepts a ``target`` or ``dev`` argument
+will automatically be parametrized over all targets specified in
+``TVM_TEST_TARGETS``.  The parametrization provides the same
+pass/fail/skipped report for each target, while allowing the test
+suite to be easily extended to cover additional targets.
+
+.. code-block:: python
+
+    # New style, implicitly parametrized to run on all
```
Comment (tkonolige) [inline docs/dev/pytest_target_parametrization.rst:116]: I thin you can remove all the new style comments given that this document is intended for how people should write targets now.

**Review by tkonolige — CHANGES_REQUESTED** [review]
Thanks for all the hard work @Lunderberg!

Given that this document is intended for people who are going to write tests, I'd first specify how they should write the tests and then talk about how it actually works. I.e. put all the parameterize_targets and requires_ stuff first.

File: docs/dev/pytest_target_parametrization.rst (hunk 1 above)
Comment (Lunderberg) [reply docs/dev/pytest_target_parametrization.rst]: Sounds good, change made.

File: docs/dev/pytest_target_parametrization.rst (hunk 2 above)
Comment (Lunderberg) [reply docs/dev/pytest_target_parametrization.rst]: Sounds good, change made.

File: docs/dev/pytest_target_parametrization.rst (hunk 3 above)
Comment (Lunderberg) [reply docs/dev/pytest_target_parametrization.rst]: Sounds good, change made.

File: docs/dev/pytest_target_parametrization.rst (hunk 4 above)
Comment (Lunderberg) [reply docs/dev/pytest_target_parametrization.rst]: Sounds good, change made.

**Comment by Lunderberg** [comment]
@tkonolige All requested changes are made, how does it look now?

**Review by jroesch — APPROVED** [review]
When @tkonolige signs off feel free to merge @Lunderberg

**Comment by hogepodge** [comment]
Thanks @Lunderberg. All looks good to me.

**Review by leandron — APPROVED** [review]
LGTM

**Comment by leandron** [comment]
This is merged now, thanks @jroesch @hogepodge @Lunderberg and @tkonolige!

**Comment by Lunderberg** [comment]
@tkonolige and I had some follow-up discussion on moving the recommended style to the top, with the explanation of how it works and why that style is preferred underneath.  That way, it is better suited for developers writing/debugging tests, who would primarily want to know what to write, but don't want to get bogged down in the how/why.  Since this one is merged, I'll add it in an additional PR.

**Pull request merged**